### PR TITLE
[testing] draft cases for private generics

### LINKED ISSUE #265: [Testing] Test cases for runtime private_generics verification.
We should write test cases for runtime verification of the `#[private_generics(T)`. These test cases can be divided into two categories: successful verification and failed verification. Additionally, multiple test scenarios should be constructed for each category.

You can refer to these issues:

- https://github.com/rooch-network/rooch/issues/252
- https://github.com/rooch-network/rooch/issues/210
- https://github.com/rooch-network/rooch/issues/235

### PR DESCRIPTION
**Pull request by feliciss**
Attempt to resolve #265

### CODE REVIEW

**Comment by steelgeek091** [comment]
The error message should be explicitly displayed in the exp file.

**Comment by feliciss** [comment]
> The error message should be explicitly displayed in the exp file.

Noted. I'll try to add them.

File: crates/rooch-framework-tests/tests/cases/private_generics/verify_private_generics_in_primitive_types.exp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+50 │         bar<vector<u64>>();
+   │         ^^^^^^^^^^^^^^^^^^
+
+error: resource type "Vector" in function "0x42::test::bar" not defined in current module or not allowed
+   ┌─ /tmp/tempfile:41:9
+   │
+41 │         bar<vector<u8>>();
+   │         ^^^^^^^^^^^^^^^^^
+
+status EXECUTED
```
Comment (jolestar) [inline crates/rooch-framework-tests/tests/cases/private_generics/verify_private_generics_in_primitive_types.exp:118]: If the verifier raises an error, the module should not publish success, and the status is not `EXECUTED` @steelgeek091

Comment (steelgeek091) [reply crates/rooch-framework-tests/tests/cases/private_generics/verify_private_generics_in_primitive_types.exp:142]: Fxied in PR https://github.com/rooch-network/rooch/pull/345

File: crates/rooch-framework-tests/tests/cases/private_generics/verify_private_generics_outside_module.move [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        v: u128,
+    }
+
+    #[private_generics(V)]
+    fun publish_bar<V: drop>(s: &signer) {
+        move_to<Bar>(s, Bar { v: 100 })
+    }
+
+    public fun invoke_publish_bar(s: &signer) {
+        publish_bar<Foo>(s);
```
Comment (steelgeek091) [inline crates/rooch-framework-tests/tests/cases/private_generics/verify_private_generics_outside_module.move:29]: After merging PR https://github.com/rooch-network/rooch/pull/345, some modifications need to be made to the test cases.

Comment (feliciss) [reply crates/rooch-framework-tests/tests/cases/private_generics/verify_private_generics_outside_module.move:29]: Fixed.

File: crates/rooch-framework-tests/tests/cases/private_generics/verify_private_generics_in_primitive_types.exp (hunk 1 above)
Comment (feliciss) [reply crates/rooch-framework-tests/tests/cases/private_generics/verify_private_generics_in_primitive_types.exp]: Fixed this PR.

**Review by steelgeek091 — COMMENTED** [review]
The relevant detection Rust code for private_generics has been updated. Please make the necessary modifications and resubmit.

**Comment by feliciss** [comment]
> The relevant detection Rust code for private_generics has been updated. Please make the necessary modifications and resubmit.

I think you should change the decoration of datatest_stable::harness!?

When changing the line

```rs
use rooch_integration_test_runner::run_test;
```

to 

```rs
use rooch_integration_test_runner::run_integration_test_with_extended_check;
```

and use it in test cases in replacement to run_test:

```rs
datatest_stable::harness!(run_integration_test_with_extended_check, "integration-tests", r".*\.move");
```

There would be errors stating that

```zsh
error[E0308]: mismatched types
  --> crates/rooch-integration-test-runner/tests/tests.rs:3:1
   |
3  | datatest_stable::harness!(run_integration_test_with_extended_check, "integration-tests", r".*\.move");
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | incorrect number of function parameters
   | arguments to this function are incorrect
   |
   = note: expected fn pointer `for<'a> fn(&'a Path) -> Result<_, _>`
                 found fn item `for<'a, 'b, 'c> fn(&'a Path, ..., ...) -> ... {run_integration_test_with_extended_check}`
note: associated function defined here
  --> runner.rs:30:12
   |
30 |     pub fn new(
   |            ^^^
   = note: this error originates in the macro `datatest_stable::harness` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rooch-integration-test-runner` due to previous error
```

I may originate from adding additional params to function `run_integration_test_with_extended_check` but no changes leave in `datatest_stable::harness!` function?

**Comment by steelgeek091** [comment]
> > The relevant detection Rust code for private_generics has been updated. Please make the necessary modifications and resubmit.
> 
> I think you should change the decoration of datatest_stable::harness!?
> 
> When changing the line
> 
> ```rust
> use rooch_integration_test_runner::run_test;
> ```
> 
> to
> 
> ```rust
> use rooch_integration_test_runner::run_integration_test_with_extended_check;
> ```
> 
> and use it in test cases in replacement to run_test:
> 
> ```rust
> datatest_stable::harness!(run_integration_test_with_extended_check, "integration-tests", r".*\.move");
> ```
> 
> There would be errors stating that
> 
> ```shell
> error[E0308]: mismatched types
>   --> crates/rooch-integration-test-runner/tests/tests.rs:3:1
>    |
> 3  | datatest_stable::harness!(run_integration_test_with_extended_check, "integration-tests", r".*\.move");
>    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>    | |
>    | incorrect number of function parameters
>    | arguments to this function are incorrect
>    |
>    = note: expected fn pointer `for<'a> fn(&'a Path) -> Result<_, _>`
>                  found fn item `for<'a, 'b, 'c> fn(&'a Path, ..., ...) -> ... {run_integration_test_with_extended_check}`
> note: associated function defined here
>   --> runner.rs:30:12
>    |
> 30 |     pub fn new(
>    |            ^^^
>    = note: this error originates in the macro `datatest_stable::harness` (in Nightly builds, run with -Z macro-backtrace for more info)
> 
> For more information about this error, try `rustc --explain E0308`.
> error: could not compile `rooch-integration-test-runner` due to previous error
> ```
> 
> It may originate from adding additional params to function `run_integration_test_with_extended_check` but no changes leave in `datatest_stable::harness!` function?

The situation is the same for `crates/rooch-integration-test-runner` and `crates/rooch-framework-tests`. However, when executing `rooch move integration-test` under the move package, the situation is different from the previous two and requires special handling.

**Comment by feliciss** [comment]
> > > The relevant detection Rust code for private_generics has been updated. Please make the necessary modifications and resubmit.
> > 
> > 
> > I think you should change the decoration of datatest_stable::harness!?
> > When changing the line
> > ```rust
> > use rooch_integration_test_runner::run_test;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to
> > ```rust
> > use rooch_integration_test_runner::run_integration_test_with_extended_check;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and use it in test cases in replacement to run_test:
> > ```rust
> > datatest_stable::harness!(run_integration_test_with_extended_check, "integration-tests", r".*\.move");
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There would be errors stating that
> > ```shell
> > error[E0308]: mismatched types
> >   --> crates/rooch-integration-test-runner/tests/tests.rs:3:1
> >    |
> > 3  | datatest_stable::harness!(run_integration_test_with_extended_check, "integration-tests", r".*\.move");
> >    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >    | |
> >    | incorrect number of function parameters
> >    | arguments to this function are incorrect
> >    |
> >    = note: expected fn pointer `for<'a> fn(&'a Path) -> Result<_, _>`
> >                  found fn item `for<'a, 'b, 'c> fn(&'a Path, ..., ...) -> ... {run_integration_test_with_extended_check}`
> > note: associated function defined here
> >   --> runner.rs:30:12
> >    |
> > 30 |     pub fn new(
> >    |            ^^^
> >    = note: this error originates in the macro `datatest_stable::harness` (in Nightly builds, run with -Z macro-backtrace for more info)
> > 
> > For more information about this error, try `rustc --explain E0308`.
> > error: could not compile `rooch-integration-test-runner` due to previous error
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It may originate from adding additional params to function `run_integration_test_with_extended_check` but no changes leave in `datatest_stable::harness!` function?
> 
> The situation is the same for `crates/rooch-integration-test-runner` and `crates/rooch-framework-tests`. However, when executing `rooch move integration-test` under the move package, the situation is different from the previous two and requires special handling.

Got it. Code in this PR are under directories `crates/rooch-integration-test-runner` and `crates/rooch-framework-tests` so it doesn't need special handlings.

I will remain them unchanged and request your review.

**Review by steelgeek091 — APPROVED** [review]

**Pull request merged**